hi-ml-cpath/testhisto/testhisto/models/test_deepmil.py

```diff
@@ -720,3 +731,81 @@ def test_ssl_containers_default_checkpoint(container_type: BaseMIL) -> None:
+) -> None:
+
+    def _get_loss(module: DeepMILModule, batch: Dict, batch_idx: int) -> Tensor:
+        loss = module.training_step(batch, batch_idx)[ResultsKey.LOSS]
+        loss.retain_grad()
+        loss.backward()
+        assert loss.grad is not None
+        assert loss.shape == (1, 1)
+        assert isinstance(loss, Tensor)
+        return loss
+
+    limit = 2
+
+    for batch_idx, batch in enumerate(dataloader):
+        if encoder_type == SwinTransformer_NoPreproc.__name__:
+            batch[SlideKey.IMAGE] = [torch.randint(0, 255, (4, 3, 224, 224), dtype=torch.uint8) / 255.] * 2
+        loss_ckpt_enc = _get_loss(model_ckpt_enc, batch, batch_idx)
+        loss_no_ckpt_enc = _get_loss(model_no_ckpt_enc, batch, batch_idx)
+        assert torch.allclose(loss_ckpt_enc, loss_no_ckpt_enc, atol=1e-4)
+        assert torch.allclose(loss_ckpt_enc.grad, loss_no_ckpt_enc.grad, atol=1e-4)
+        if batch_idx == limit:
+            break
+
+
+@pytest.mark.parametrize(
+    "encoder_type, feature_dim",
+    [(Resnet18.__name__, 512), (Resnet50.__name__, 2048), (SwinTransformer_NoPreproc.__name__, 768)],
+)
+def test_encoder_checkpointning(encoder_type: str, feature_dim: int, mock_panda_tiles_root_dir: Path) -> None:
+    container = MockDeepSMILETilesPanda(tmp_path=mock_panda_tiles_root_dir)
+    container.setup()
+
+    data_module = container.get_data_module()
+    train_dataloader = data_module.train_dataloader()
+    val_dataloader = data_module.val_dataloader()
+
+    container.encoder_type = encoder_type
+
+    model = container.create_model()
+    model.trainer = MagicMock(world_size=1)  # type: ignore
+    model.outputs_handler = MagicMock()
+    model.log = MagicMock()  # type: ignore
+
+    model_no_ckpt_enc = model
+    model_ckpt_enc = deepcopy(model)
+    model_ckpt_enc.encoder_params.checkpoint_encoder = True
+    model_ckpt_enc.encoder.checkpoint_activations = True
+
+    # 1. Compare the loss and gradients of the encoder with and without checkpointing
+    validate_loss_with_activations_checkpointing(train_dataloader, model_ckpt_enc, model_no_ckpt_enc, encoder_type)
+
+    if encoder_type != SwinTransformer_NoPreproc.__name__:  # SwinT requires images of 224 input size, mock tiles are 28
+        # 2. Train the model with and without checkpointing and compare the encoder parameters
+        trainer_no_ckpt = Trainer(max_epochs=1, limit_train_batches=2, limit_val_batches=2, limit_test_batches=2)
+        trainer_no_ckpt.fit(model_no_ckpt_enc, train_dataloader, val_dataloader)
+
+        trainer_ckpt = Trainer(max_epochs=1, limit_train_batches=2, limit_val_batches=2, limit_test_batches=2)
+        trainer_ckpt.fit(model_ckpt_enc, train_dataloader, val_dataloader)
+
+        for param_no_ckpt, param in zip(model_no_ckpt_enc.encoder.parameters(), model_ckpt_enc.encoder.parameters()):
+            assert torch.allclose(param_no_ckpt, param, atol=1e-4)
```

Comment:
can we parametrize bn_momentum? do we expect to see this with different bn_momentum values?

Reply:
good point, added it thanks!